FXL6408 A1 (core functionalities implemented)

Ilagas opened the pull request:
created this branch so i can add only the relative files to the pull request

abelkarapatis commented on gpio_fxl6408.c:
```diff
@@ -0,0 +1,486 @@
+#define DT_DRV_COMPAT fcs_fxl6408
+
+/* Register definitions */
+#define REG_DEVICE_ID_CTRL		0x01
```
Would be nice to have indentation a bit more aligned

abelkarapatis commented on gpio_fxl6408.c:
```diff
@@ -0,0 +1,486 @@
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	const struct device *i2c_master = drv_data->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	uint8_t port_data;
```
Make sure you always initialize variables

`uint8_t data = 0;`

abelNoelleC commented:
Seems like you put the files in the rootdirectory. If that's correct, can you please move them back to the 'drivers/gpio' (if I remember correctly) folder where they were before?

Edit: what I mean is that it would have been fine to use the branch you already had, for this PR. That way we can see what the actual changes are, and now the git system can't properly show the changes (like you can see in the CMakeLists.txt)

Ilagas committed:
drivers: fxl6408: FXL6408 pull-request base

Basis for the FXL6408. It is in working state but necesary to
clean up some more.

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

abelNoelleC commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -35,9 +35,7 @@ zephyr_library_sources_ifdef(CONFIG_GPIO_XLNX_AXI   gpio_xlnx_axi.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_NPCX       gpio_npcx.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_EMUL       gpio_emul.c)
 zephyr_library_sources_ifdef(CONFIG_GPIO_PSOC6      gpio_psoc6.c)
-zephyr_library_sources_ifdef(CONFIG_GPIO_PCAL6408A  gpio_pcal6408a.c)
```
what happened to these lines? are the branches not branched from the same point/commit?

abelNoelleC commented on drivers/gpio/Kconfig:
```diff
@@ -85,8 +85,6 @@ source "drivers/gpio/Kconfig.emul"
 
 source "drivers/gpio/Kconfig.psoc6"
 
-source "drivers/gpio/Kconfig.pcal6408a"
```
same question here: what happened to these lines? are the branches not branched from the same point/commit?

abelNoelleC commented on samples/FXL_I2C/boards/fxl.dtsi:
```diff
@@ -0,0 +1,17 @@
+		ngpios = <8>;
+        #gpio-cells = <2>;
+        gpio-controller;
+        //vin-supply = <&vdd_pwr>;
```
please remove code that isnt relevant for the PR

abelNoelleC commented on samples/FXL_I2C/boards/stm32l1_disco.overlay:
```diff
@@ -0,0 +1,33 @@
+	status = "okay";
+	pinctrl-0 = <&i2c1_scl_pb6 &i2c1_sda_pb7>;
+//	pinctrl-0 = <&i2c0_gpio44>;
+	//compatible = "st,";
```
please remove code that isnt relevant for the PR

abelNoelleC commented on samples/FXL_I2C/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_LOG=y
```
Please check which ones of these are already included in the board/SoC Kconfig, and separate them like this (or similar):

`### required for driver but selected by board/SoC already ###`
`CONFIG_GPIO=y`
`CONFIG_SERIAL=y`

`### required for driver and not selected by board/SoC ###`
`CONFIG_MODEM=y`
`CONFIG_MODEM_UBLOX_SARA_N310=y`

abelNoelleC commented on samples/FXL_I2C/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+sample:
```
This file is used for the testing framework, will have to check when doing the zephyrproject PR if this is okay. For now it's okay like this

abelNoelleC commented on samples/FXL_I2C/src/main.c:
```diff
@@ -0,0 +1,70 @@
+
+void main(void)
+{
+	gpio_dev = device_get_binding("FXL6408");
```
binding will probably have to be gotten in a different way, most likely something similar to [[this](https://github.com/abelsensors/zephyr/pull/2/files#diff-f41f6a6439299463fc93f46da75c3739400e9935220485ac67013cbb0fe4d2a5R25)]

abelNoelleC reviewed (commented):
Looking good! 
Please also create a README file describing the functionality and how the driver can be used (with examples in the sample application). Also please share some of the test results / a demonstration in the PR description.

abelNoelleC commented:
Nice! A few things still:
- Please either (respond to and) 'resolve' the comments that haven't automatically been resolved by github through the code changes
- I still see a merge conflict in CMakeLists, can you solve that?
- You still haven't added a README file describing the current functionalities and how the driver is used by the sample application, please add that
- I'm still missing some comments in the driver, describing globally what's happening (obviously doesn't have to be done for every line, but please see if there's some groups of lines that you can summarize and create a comment for). 

Almost there!

abelkarapatis commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,436 @@
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	uint8_t reg_out;
```
@Ilagas  Please make sure you initialize the variables

abelkarapatis commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright (c) 2020 Abel Sensors
```
Isn't it 2021 ? :p

Ilagas committed:
drivers: gpio: updated fxl6408 license to correct date

updated license and added README file

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

Ilagas committed:
docs: gpio: Added comments to code

Added comments to make code more clear. Could add more
but dont see where.

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

Ilagas committed:
drivers: gpio: clean up

Removed some style errors

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

Ilagas committed:
drivers: gpio: still more style errors

Still more style errors

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

Ilagas committed:
drivers: gpio: need more space

Needed 1 more space in the fxl7408 file

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

abelkarapatis commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,445 @@
+{
+	struct gpio_fxl6408_drv_data *const drv_data =
+		(struct gpio_fxl6408_drv_data *const)dev->data;
+	int ret = read_port_regs(dev, REG_INPUT_VALUE,
```
I guess you could rewrite this as 
```
return read_port_regs(dev, REG_INPUT_VALUE,
                 &drv_data->reg_cache.input, buf);
```

Ilagas replied:
compared to the PCA95, i changed 
`int read_port_regs(const struct device *dev, uint8_t reg, uint8_t *cache, uint8_t *buf )`
to 
`int read_port_regs(const struct device *dev, uint8_t reg, uint8_t *cache )`
because cache and buf return the same value and I thought this way would be cleaner. I could be wrong though.

abelkarapatis replied:
Didn't notice that. That is fine :)

abelkarapatis commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,445 @@
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW)) {
+			reg_out &= ~BIT(pin);
+		}
+		reg_dir |= BIT(pin);
```
Could you explain why this is different from the pca95xxx.c ?

FXL 
```
		reg_dir |= BIT(pin);
	} else {
		reg_dir &= ~BIT(pin);
	}
```
PCA
```
		reg_dir &= ~BIT(pin);
	} else {
		reg_dir |= BIT(pin);
	}
```

Ilagas replied:
With the PCA95, writing a 1 (over I2C) to a pin direction means it is configured as input. With the FXL, writing a 1 means it is configured as output.

abelkarapatis commented on drivers/gpio/gpio_fxl6408.c:
```diff
@@ -0,0 +1,445 @@
+ */
+
+#include <zephyr.h>
+#include <sys/printk.h>
```
sys/printk.h is not needed anymore, right ?

abelkarapatis replied:
Please also check if the following are needed

- dt-bindings/gpio/gpio.h
- soc.h
- zephyr.h

Ilagas committed:
drivers: gpio: removed unused headers

removed al headers except 4 which were actually used

Signed-off-by: Imre Lagas <imre_lagas@hotmail.com>

abelkarapatis commented:
Everything looks! Approved :)
